docs/man/serial/en/serial.1.md
% SERIAL(1)
% Naohiro CHIKAMATSU <n.chika156@gmail.com>
% September 2021

# NAME

serial – rename the file name to the name with a serial number.

# SYNOPSIS

**serial** [OPTIONS] DIRECTORY_PATH

# DESCRIPTION
**serial** is a CLI command that renames files under any directory to  
the format user specified file name with serial number.  
serial can specify the destination directory of the renamed file. Also,  
if you want to keep the original file, you can copy the file instead of  
renaming the file.

# EXAMPLES
**Rename the file name to the name with a serial number at current directory.**  

    $ ls  
      a.txt  b.txt  c.txt  
    $ serial --name demo  .  
      Rename a.txt to 1_demo.txt  
      Rename b.txt to 2_demo.txt  
      Rename c.txt to 3_demo.txt

**Copy & Rename the file at specified directory.**  

    $ serial -s -k -n ../../dir/demo .  
      Copy a.txt to ../../dir/demo_0.txt  
      Copy b.txt to ../../dir/demo_1.txt  
      Copy c.txt to ../../dir/demo_2.txt

# OPTIONS
**-d**, **--dry-run**
:   Output the file renaming result to standard output (do not update the file).

**-f**, **--force**
:   Forcibly overwrite and save even if a file with the same name exists.

**-h**, **--help**
:   Show help message.

**-k**, **--keep**
:   Keep the file before renaming (not rename, do copy).

**-n new_filename**, **--name=new_filename**
:   Base file name with/without directory path (assign a serial number to this file name).

**-p**, **--prefix**
:   Add a serial number to the beginning of the file name(default).

**-s**, **--suffix**
:   Add a serial number to the end of the file name.

**-v**, **--version**
:   Show serial command version.

# EXIT VALUES
**0**
:   Success

**1**
:   There is an error in the argument of the serial command.

# BUGS
See GitHub Issues: https://github.com/nao1215/mimixbox/issues

# LICENSE
The MimixBox project is licensed under the terms of the MIT license and Apache License 2.0.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
## [0.13.0] - 2021-11-20
### Added
 - cowsay command. The process of enclosing the message in a frame is incomplete
 - docker target to Makefile. This target was created to test Mimixbox inside Docker.
 ### Changed
 - Fixed a bug in the cat command. This bug occurs when a standard input is accepted more than once and then an empty enter is received on the next input. In the correct behavior, it is correct to output a blank line, but since the previous input value has been saved, the previous input value is output.
## [0.12.1] - 2021-11-20
### Added
 - base64 command.
 - cp command.
 - nl command.
 - -n option for cat command
### Changed
 - Reduce mimixbox binary size by compile option (7.5MB --> 5.4MB)
 - cat/tac command to receive data from standard input when the argument is "-".
## [0.9.1] - 2021-11-19
### Added
 - basename command.
 - sleep command.
 - tac command.
### Changed
 - The library that was open to the public(pkg) to the internal library.
## [0.6.0] - 2021-11-18
### Added
 - mkfifo command.
 - rm command.
 - rmdir command.
### Changed
 - cat/tac commands does not output a help message even if there is no argument.
### Fixed
 - Bug that misjudged applet arguments as mimixbox arguments.
## [0.3.0] - 2021-11-17
### Added
 - touch command.
 - licenses target to Makefile.

## [0.2.0] - 2021-11-17
### Added
 - mv command.
## [0.1.1] - 2021-11-16
### Added
 - ischroot command.
 - mkJailForDebianFamily.sh that make rootfs for testing chroot/ischroot command.
 - jail target to Makefile. Only work for Debian-based distribution.
### Changed
 - All applet returns error and exit code for main code.

## [0.0.1] - 2021-11-14
### Added
 - mimixbox project.

internal/applets/shellutils/serial/serial.go
//
//  mimixbox/internal/applets/shellutils/serial/serial.go
//
// Copyright 2021 Naohiro CHIKAMATSU
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package serial

import (
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"

	mb "github.com/nao1215/mimixbox/internal/lib"

	"github.com/jessevdk/go-flags"
)

const cmdName string = "serial"

var osExit = os.Exit

const version = "1.0.1"

// Exit code
const (
	ExitSuccess int = iota // 0
	ExitFailuer
)

type options struct {
	DryRun  bool   `short:"d" long:"dry-run" description:"Output the file renaming result to standard output (do not update the file)"`
	Force   bool   `short:"f" long:"force" description:"Forcibly overwrite and save even if a file with the same name exists"`
	Keep    bool   `short:"k" long:"keep" description:"Keep the file before renaming"`
	Name    string `short:"n" long:"name" value-name:"<name>" description:"Base file name with/without directory path (assign a serial number to this file name)"`
	Prefix  bool   `short:"p" long:"prefix" description:"Add a serial number to the beginning of the file name(default)"`
	Suffix  bool   `short:"s" long:"suffix" description:"Add a serial number to the end of the file name"`
	Version bool   `short:"v" long:"version" description:"Show serial command version"`
}

func Run() (int, error) {
	var opts options
	var args = parseArgs(&opts)
	var dirPath = args[0]

	if !mb.Exists(dirPath) {
		err := fmt.Errorf("%s doesn't exist.", dirPath)
		return ExitFailuer, err
	}

	var files = getFilePathsInDir(dirPath)
	if len(files) == 0 {
		err := fmt.Errorf("No files in %s directory.", dirPath)
		return ExitFailuer, err
	}

	newFileNames := newNames(opts, files)
	dieIfExistSameNameFile(opts.Force, newFileNames)
	makeDirIfNeeded(newFileNames[files[0]])

	if opts.Keep {
		copy(newFileNames, opts.DryRun)
	} else {
		rename(newFileNames, opts.DryRun)
	}
	return ExitSuccess, nil
}

func rename(newFileNames map[string]string, dryRun bool) {
	keys := make([]string, 0, len(newFileNames))
	for k := range newFileNames {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	for _, org := range keys {
		fmt.Printf("Rename %s to %s\n", org, newFileNames[org])
		if dryRun {
			continue
		}
		if err := os.Rename(org, newFileNames[org]); err != nil {
			fmt.Fprintf(os.Stderr, "Can't rename %s to %s\n", org, newFileNames[org])
			osExit(1)
		}
	}
}

func copy(newFileNames map[string]string, dryRun bool) {
	var dest string
	keys := make([]string, 0, len(newFileNames))

	for k := range newFileNames {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	for _, org := range keys {
		dest = newFileNames[org]
		fmt.Printf("Copy %s to %s\n", org, dest)
		if dryRun {
			continue
		}
		// In the case of renaming, even the same file name can be overwritten.
		// On the other hand, in the case of copying, an error will occur
		// if serial command try to overwrite with the same file name.
		if org == dest {
			continue
		}

		// If this function is running, it will force the file to be overwritten.
		// If there is the file with the same name in the copy destination,
		// delete it before copy the file.
		if mb.Exists(dest) {
			if err := os.Remove(dest); err != nil {
				fmt.Fprintf(os.Stderr, "Can't copy %s to %s\n", org, dest)
				osExit(ExitFailuer)
			}
		}

		if err := os.Link(org, dest); err != nil {
			fmt.Fprintf(os.Stderr, "Can't copy %s to %s\n", org, dest)
			osExit(ExitFailuer)
		}
	}
}

func parseArgs(opts *options) []string {
	p := initParser(opts)

	args, err := p.Parse()
	if err != nil {
		osExit(ExitFailuer)
	}

	if opts.Version {
		showVersion()
		osExit(ExitSuccess)
	}

	if len(opts.Name) != 0 && !existFilenameInPath(opts.Name) {
		showHelp(p)
		osExit(ExitFailuer)
	}

	if !isValidArgNr(args) {
		showHelp(p)
		osExit(ExitFailuer)
	}

	return args
}

func initParser(opts *options) *flags.Parser {
	parser := flags.NewParser(opts, flags.Default)
	parser.Name = cmdName
	parser.Usage = "[OPTIONS] DIRECTORY_PATH"

	return parser
}

func isValidArgNr(args []string) bool {
	return len(args) == 1
}

func showVersion() {
	fmt.Printf("serial version %s\n", version)
}

func showHelp(p *flags.Parser) {
	fmt.Printf("serial command rename the file name to the name with a serial number.\n\n")
	p.WriteHelp(os.Stdout)
}

func existFilenameInPath(path string) bool {

	return !strings.HasSuffix(path, "/")
}

// getFilePathsInDir returns the paths of the file in the directory.
func getFilePathsInDir(dir string) []string {
	files, err := ioutil.ReadDir(dir)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Can't get file list.")
		osExit(ExitFailuer)
	}

	var path string
	var paths []string
	for _, file := range files {
		path = filepath.Join(dir, file.Name())
		if mb.IsFile(path) && !mb.IsHiddenFile(path) {
			paths = append(paths, filepath.Clean(path))
		}
	}
	sort.Strings(paths)
	return paths
}

func newNames(opts options, path []string) map[string]string {
	newNames := make(map[string]string)
	destDir := filepath.Dir(opts.Name)

	var fileName string
	var format string
	// TODO: Refactor for a simpler implementation
	for i, file := range path {
		ext := filepath.Ext(file)

		if len(opts.Name) == 0 {
			format = fileNameFormat(opts.Prefix, opts.Suffix, mb.BaseNameWithoutExt(file), len(path))
		} else {
			format = fileNameFormat(opts.Prefix, opts.Suffix, opts.Name, len(path))
		}

		if opts.Prefix && !opts.Suffix {
			fileName = fmt.Sprintf(format, i, ext)
		} else {
			fileName = fmt.Sprintf(format, i, ext)
		}

		if destDir == "." {
			newNames[file] = filepath.Clean(fileName)
		} else {
			newNames[file] = filepath.Clean(destDir + "/" + fileName)
		}
	}
	return newNames
}

func fileNameFormat(prefix bool, suffix bool, name string, totalFileNr int) string {
	baseName := filepath.Base(name)
	serial := "%0" + strconv.Itoa(len(strconv.Itoa(totalFileNr))) + "d"
	ext := "%s"

	// Default format（e.x.：%03d%s%s → 001_test.txt）
	format := serial + "_" + baseName + ext
	if !prefix && suffix {
		format = baseName + "_" + serial + ext
	}
	return format
}

func dieIfExistSameNameFile(force bool, fileNames map[string]string) {
	if force {
		return
	}

	for _, file := range fileNames {
		if mb.Exists(file) {
			fmt.Fprintf(os.Stderr, "%s (file name which is after renaming) is already exists.\n", file)
			fmt.Fprintf(os.Stderr, "Renaming may erase the contents of the file. ")
			fmt.Fprintf(os.Stderr, "So, nothing to do.\n")
			osExit(ExitFailuer)
		}
	}
}

func makeDirIfNeeded(filePath string) {
	dirPath := filepath.Dir(filePath)

	if mb.Exists(dirPath) {
		return
	}

	if err := os.MkdirAll(dirPath, 0755); err != nil {
		fmt.Fprintf(os.Stderr, "Can't make %s directory\n", dirPath)
		osExit(ExitFailuer)
	}
}
